### PR TITLE
feat: new ledger channel algorithm

### PR DESCRIPTION
**Pull request by tomclose**
Adds a new ledger channel algorithm

## Background

The ledger funding algorithm in current master (roughly) works as follows:

- When an objective wants a channel to added to / removed from a ledger channel it creates a `LedgerRequest`. The request stores the `channelId`, `ledgerId`, and whether it's a fund/defund.
- After all objectives have been cranked, the `LedgerManager` processes the ledger request queue.
- In order to decide what update to propose, the participants exchange *ledger proposals.* Roughly, both participants send each other the list of changes they're willing to make at this point, and the actual change is computed by taking the intersection.
- A ledger update is then proposed containing this intersection, which should be accepted by each party.

As far as I'm aware, there isn't high confidence that the current implementation is robust. ~~There have been some efforts to show that it is, which haven't yet been successful.~~

## Changes made / motivations

This work suggests a new ledger funding algorithm. It makes the following changes:

1. No ledger proposals are required, instead the ledger states themselves are used to communicate proposals. This allows us to remove an object from the store and wire protocol, and should improve robustness, as we can lean on the existing mechanisms for syncing states between participants. We still obtain agreement in a single round-trip. 
2. Adds an efficient and robust mechanism for handling the case where a protocol wants to cancel funding it has requested. In short, a protocol can cancel a funding request with a defund request. The ledger manager will cancel the initial fund request if it can, or defund if it is too late to cancel.
3. Ledger requests now hold the diff they want to apply to the ledger, instead of the just the channelId. This decouples the ledger manager from the state of the channel, which should make it easier to add more exotic ledger operations (e.g. partial withdrawals) in the future. It also slightly more efficient, as it means the ledger manager doesn't need to query the channel.
4. It adds a "missedOpportunityCount" to the ledger request, which tracks how many time the request could have been included in a new agreed state but wasn't. In the future the wallet can use this information to detect ledger updates that have likely stalled (e.g. because the opponent has abandoned/lost the objective), and take appropriate action.

A detailed description of the algorithm is included in the [comments in the code](https://github.com/statechannels/statechannels/blob/tc/new-ledger-algorithm/packages/server-wallet/src/protocols/ledger-manager.ts#L11-L68). 

## Alternative approaches considered

I spent some time considering the "ledger updates as objectives" proposal. My notes are [here](https://www.notion.so/Ledger-Channels-d5af6ed79b7e4d4eb3e3511b7b8f8b18). In short:

- In the current one-objective-per-channel architecture, ledger updates as objectives breaks down pretty quickly, unless you stick to the case where a ledger is used to fund a single set of channels that are funded / defunded at the same time.
- There are no alternative written proposals for architectures that don't use one-objective-per-channel, so I am unable to evaluate the "ledger updates as objectives" in these architectures.

In any case, my understanding is that a key motivation for "ledger updates as objectives" is that it is believed that it would be simple and would just work, and it's worth living with the downsides in the short-term for those reasons (as far as I'm aware there is no proposal outlining the approach and the motivations, so I could be wrong here). If the approach proposed here is simple and works, it dominates this alternative, as it doesn't share the limitations described above.
## How Has This Been Tested?

I have written a fairly comprehensive test suite to test the new ledger-manager.

---

### Code quality
- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [x] This change does not have an unduly wide scope
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary (can be 0)
- [x] I have added tests that prove my fix is effective or that my feature works, if necessary
### Project management
- [x] I have applied the [appropriate labels](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#b304e56734a74dfbb341b8b4b27b1c0c)
- [x] I have [linked to all relevant issues](https://help.zenhub.com/support/solutions/articles/43000010350-connecting-pull-requests-to-github-issues) (can be 0)
- [x] I have [linked to all dependent issues](https://help.zenhub.com/support/solutions/articles/43000010349-create-github-issue-dependencies) (can be 0)
- [x] I have assigned myself to this PR
- [x] I have chosen the appropriate [pipeline](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#d534e68e7edc46fe8a4cda61b2258c4e) on zenhub for the linked issue

### CODE REVIEW

**Comment by andrewgordstewart** [comment]
Is [this](https://www.notion.so/Ledger-updates-caaac6d640c2448398264d3d1c1feef3) the (only?) document inspiring this PR?

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:874]: Should we remove this test?

File: packages/server-wallet/src/protocols/ledger-funder.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/ledger-funder.ts:56]: Is this still relevant?

**Review by kerzhner — COMMENTED** [review]
I completed the first pass of the `ledger-manager`, and it looks reasonable to me! I'll take another pass tomorrow to see if I understand the tests and other code that supports the ledger-manager.

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:17]:
```suggestion
```

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:19]:
```suggestion
```

File: packages/server-wallet/src/models/ledger-request.ts
Comment (kerzhner) [reply packages/server-wallet/src/models/ledger-request.ts:9]: Maybe this is the reason we need both queued and pending statuses? https://github.com/statechannels/statechannels/blob/tc/new-ledger-algorithm/packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts#L173

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:48]: I am a big fan of this declarative test style. The tests are pleasant to read!

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:403]: Does this still apply?

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:424]: Is it true that the follower has no way of relaying this request to the leader? If I understand correctly, the leader must send a state to the follower that includes this request.

File: packages/server-wallet/src/protocols/ledger-manager.ts
Comment (tomclose) [reply packages/server-wallet/src/protocols/ledger-manager.ts:14]: I think ledger channels will likely be 2-participants for the foreseeable future (even with virtual channels, you only need two party ledgers). 

The algorithm does naturally scale to multiple parties though: the first participant is still the "leader" but you then have multiple "followers", who each either agree with what's passed to them, or remove the parts they don't like.

File: packages/server-wallet/src/protocols/ledger-manager.ts
Comment (tomclose) [reply packages/server-wallet/src/protocols/ledger-manager.ts:289]: Yes. I believe that this is currently enforced by a db constraint. This isn't an essential constraint though, and we could modify the algorithm to allow multiple fund/defund attempts per channel. It's a bit hard to imagine how that would come about though.

File: packages/server-wallet/src/protocols/ledger-manager.ts
Comment (tomclose) [reply packages/server-wallet/src/protocols/ledger-manager.ts:461]: I think the code here is wrong. Will fix.

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (tomclose) [reply packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:424]: Yes that's true. The ledger algorithm starts from a position where both participants have a queue of requests, that they want to apply to the channel. It's the objective's job to ensure that both participants have the same requests in their queue.

Comment (tomclose) [reply packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:424]: If the followers wallet notices that a request is getting a lot of missedOps, it can use that as a trigger to sync/re-send that objective.

File: packages/server-wallet/src/models/ledger-request.ts
Comment (tomclose) [reply packages/server-wallet/src/models/ledger-request.ts:9]: I think it isn't actually important for the algorithm, as it's deducible from the ledger states, and I'm pretty sure I always do this.

It might be nice to be able to communicate this information to the objective and/or user though. It's effectively "we're just waiting for your opponent to sign now". Initially I thought that objectives might behave differently according to whether the request was pending, as once a request is pending it's out of the wallets control whether it gets accepted or not, so the request can't be withdrawn. However, that can be fragile (a pending request might revert to queued; a race condition might mean that a queued request becomes pending), and the just-send-a-cancelling-defund approach handles these issues optimally and robustly.

Overall, I think I'd be inclined to leave it at this point as it (a) works, (b) isn't doing any harm and (c) potentially provides a bit of extra debugging/UX info, even though it isn't strictly necessary.

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (tomclose) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:874]: Yeah, probably. I wasn't sure what to do with this - the old ledger algorithm did a lot of extra ("pessimistic") state communication, so it could handle cases like these. The new ledger algorithm doesn't do this - it just relies on sync states being called to handle the case where things have stalled. So the new ledger algorithm fails this test, but that's probably ok?

File: packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts
Comment (tomclose) [reply packages/server-wallet/src/protocols/__test__/ledger-manager.test.ts:403]: I think we probably don't need this behaviour at this point.

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (kerzhner) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:874]: Yep, I think it is totally fine to not handle the same edge cases with the new ledger funding algorithm.

File: packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts
Comment (kerzhner) [inline packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:30]: Is the `LedgerProposals` table/model unused now? If so, should we remove code that references the model and write a db migration to remove the table?

Comment (kerzhner) [reply packages/server-wallet/src/__test-with-peers__/create-and-close-channel/ledger-funding.test.ts:30]: We can take care of this after this PR is merged. Filed https://github.com/statechannels/statechannels/issues/3371, so that we don't forget.

File: packages/server-wallet/src/models/ledger-request.ts
Comment (kerzhner) [inline packages/server-wallet/src/models/ledger-request.ts:88]:
```suggestion
  static async ledgersWithNewRequestsIds(tx: TransactionOrKnex): Promise<string[]> {
```

File: packages/server-wallet/src/models/ledger-request.ts
Comment (kerzhner) [inline packages/server-wallet/src/models/ledger-request.ts:90]: Would you be able to elaborate on how this query compares to a query that reads all ledger requests that have  `queued` or `pending` status?

Comment (tomclose) [reply packages/server-wallet/src/models/ledger-request.ts:90]: I'm using the fact that `lastSeenAgreedState` will be null if and only if the ledger request has never been involved in a crank before. These requests are therefore the ones that were added in the last round of objective cranks, and therefore these are the ledgers we need to crank.

File: packages/server-wallet/src/wallet/wallet.ts
Comment (kerzhner) [inline packages/server-wallet/src/wallet/wallet.ts:959]: This is a follow up from a previous question. These are the ledger requests with an empty `lastSeenAgreedState`. What would be the impact on this code if `getLedgersWithNewRequestsIds` returned `queued` requests?

Comment (kerzhner) [reply packages/server-wallet/src/wallet/wallet.ts:959]: I am trying to think through the following scenario. Let's say that A and B would like to ledger fund channels c and d using ledger channel L. L's initial outcome is [A, B]. Let's walk through the following sequence of events:
1. A calls `createChannel`  for c, then A calls `createChannel` for d.
2. B calls `joinChannel` for c, then B calls `joinChannel` for d.
3. B now has ledger requests for c and d.
4. A receives B's prefund signature for c. B's prefund signature on d gets delayed in transit.
5. A sends a ledger state Proposal that only includes c: [A, B, c].
6. B signs the Proposal [A, B, c] and sets the `lastSeenState` for d.
7. B's prefund signature on d finally arrives to A.
8. A receives B's signature on Proposal [A, B, c]. A now sends a second Proposal that includes d: [A, B, c, d].

I can't seem to convince myself that B's `takeActions` will process the second Proposal. I am probably missing something...

Comment (kerzhner) [reply packages/server-wallet/src/wallet/wallet.ts:959]: Maybe I am starting to understand. When B receives the Proposal of [A, B, c, d], `takeActions` should be called with `channelIds` parameter of `L`. So ` await this.store.filterChannelIdsByIsLedger(channelIds)` should result in the proposal getting processed.

Comment (tomclose) [reply packages/server-wallet/src/wallet/wallet.ts:959]: In the case above, the ledger would be included in `ledgersFromChannels` (the line below this one), and so would get processed by b. The general logic is that there are two cases where you need to crank a ledger:

1. When that ledger channel itself has received an update
2. When an objective has just added a new request to the queue

If the case where a ledger has (a) already seen all of its requests, and (b) hasn't changed, you don't need to crank again, as you'll just get the same result. So selecting `queued` requests here would lead to redundant work.

I guess there's an argument for adding a `new` status to the ledger request, to make the case of "not yet processed by a ledger crank" more explicit.

Comment (kerzhner) [reply packages/server-wallet/src/wallet/wallet.ts:959]: Thanks for the explanation!

**Review by kerzhner — APPROVED** [review]
Nicely done @tomclose!

**Pull request merged**
